---
title: "plasmaProfile"
author: "Melanie Smith"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Import Counts and Sample Data

This step imports the raw counts data and samples metadata and reports some summary information.

### Import Count Data

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# load required libraries
library(plyr)
library(dplyr)
library(tidyverse)
library(reshape2)
library(pander)
library(magrittr)
library(RColorBrewer)
library(edgeR)
library(limma)
library(pheatmap)

# Import plasma miRNA counts table from VM
countsPlasma <- read.table(("~/plasmaBcbio/plasma/final/2017-05-22_plasma/counts_mirna.tsv"), 
                sep = "\t",
                header = T, 
                row.names = "mirna")

# remove the "S#_R1_001" from the column names (eg PAC001_S1_R1_001 becomes PAC001 to match the 'samples' data.frame) in the 'counts' data.frame
names(countsPlasma) = sub("_R1_001","",names(countsPlasma))
names(countsPlasma) = sub("_S.*","",names(countsPlasma))

```

# counts QC

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

total_countsPlasma <- colSums(countsPlasma) %>%
   as.data.frame() %>%
   tibble::rownames_to_column() %>%
   set_colnames(c("samplename","totalReads"))

#pdf("plasmaTotalReads.pdf")
ggplot(data = total_countsPlasma, aes(x = samplename, y = totalReads, fill = samples$group)) +
   geom_bar(stat = "identity") +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
   labs(x = "Sample Name", y = "Total Reads") + 
   geom_hline(yintercept=5.0e+06,linetype = "dashed") +
   scale_fill_discrete(name = "Trimester")
#dev.off()


# WriteXLS("total_countsPlasma", 
#             ExcelFileName = "total_countsPlasma.xls", 
#             row.names = TRUE)

head_countsPlasma <- countsPlasma %>%
   tibble::rownames_to_column() %>%
   arrange(desc(PAC006)) %>%
   filter(., PAC006 > 100000)

# WriteXLS("head_countsPlasma", 
#             ExcelFileName = "head_countsPlasma.xls", 
#             row.names = TRUE)

totalReads <- colSums(t(countsPlasma)) %>%
   as.data.frame() %>%
   set_colnames("totalReads_per_miR") %>%
   arrange(PAC006)

ggplot(data = total_countsPlasma, aes(x = samplename, y = totalReads, fill = samples$group)) +
   geom_bar(stat = "identity") +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
   labs(x = "Sample Name", y = "Total Reads") + 
   geom_hline(yintercept=5.0e+06,linetype = "dashed") +
   scale_fill_discrete(name = "Trimester")


```

### Import Sample Meta Data

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}
# import the samples data from VM 
samples <- readr::read_delim("~/placenta_miRNA.tsv", delim = "\t") %>% 
  dplyr::select(-starts_with("X"))

samples <- samples[,c(1, 12:16)] %>%
  as.data.frame() %>%
  magrittr::set_colnames(c("samplename", "maternalAge", "gestation", "BMI", "smoker", "ethnicity")) %>% 
  mutate(samplename = paste0("PAC",stringr::str_pad(samplename, width = 3, pad = "0")))

# Remove the "+" from column 4 "gestation" nb: escape special character using \\
samples$gestation <- gsub("\\+", "", samples$gestation)

# reassign object classes
samples$gestation <- as.integer(samples$gestation)

# add new column named "group" into samples data.frame 
samples$group <- rep(NA, 48)

# add new group 'Trimester 1 or 2' factor to each row based on gestation
# Trimester 1 = 6-13 wks, Trimester 2 = 14-26 wks
samples$group <- ifelse(samples$gestation <= 13, "Trimester_1", 
                        ifelse(samples$gestation >= 14, "Trimester_2", NA))

```

### Reassign object classes

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

samples$group <- as.factor(samples$group) 

samples <- as.data.frame(samples)

```

### Import and add fetal sex data to Sample Meta Data

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# Add fetal sex to samples df
fetalSex <- read.csv("~/Bcbioplasma/fetalSex.csv", header = TRUE) %>%
  dplyr::select(-matches("gestation")) %>%
  mutate(samplename = paste0("PAC", stringr::str_pad(samplename, width = 3, pad = "0")))

fetalSex$Sex <- ifelse(fetalSex$fetalSex == "XX" , "Female", 
                       ifelse(fetalSex$fetalSex == "XY", "Male", NA))

samples <- left_join(samples, fetalSex, by = "samplename")

```

### Summary Data

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# summary tables
pandoc.table(samples,
             caption = "Meta Data for all plasma samples",
             split.table = Inf)

summary(samples$group)
summary(samples$fetalSex)


lapply(samples[, c(2:8)], function(x) {
    if (is.numeric(x)) return(summary(x))
    if (is.integer(x)) return(summary(x))
    if (is.factor(x)) return(table(x))
    if (is.character(x)) return(table(x))
})

```

### Plot to visualise the pre filtered, unique miRNA population between samples in Maternal Plasma

From the raw counts we can see that is considerable difference in the number of unique miRNAs identified between samples.

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# set ggplot2 theme
theme_set(theme_bw())

# count the number of 'non-zero' (ie expressed) miRNAs for each sample before filtering
nonzero <- function(x) sum(x != 0)

rawNonZero <- plyr::numcolwise(nonzero)(countsPlasma) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  set_colnames(c("samplename", "rawmiRNA")) %>%
  as.data.frame()

# produce plot of # unique miRNA per sample - fill by trimester
#png("rawmiRNAPlasma.png")
ggplot(data = rawNonZero, aes(x = samplename, y = rawmiRNA, fill = samples$group)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "Sample Name", y = "Count of miRNAs", title = "Count of pre-filtered miRNAs in Plasma") + 
  geom_hline(yintercept=1571,linetype = "dashed") +
  geom_text(aes(20, 1600, label = "1571 Total miRNA"), size = 4, colour = "black") +
  scale_fill_discrete(name = "Trimester")
#dev.off()


# produce plot of # unique miRNA per sample - no fill
#pdf("rawmiRNAPlasma.pdf")
ggplot(data = rawNonZero, aes(x = samplename, y = rawmiRNA)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "Sample Name", y = "Count of miRNAs", title = "Count of pre-filtered miRNAs in Plasma") +
  theme_bw()
#dev.off()

```

# QC miRNA populations

```{r}

library(ggplot2)

nba <- read.csv("http://datasets.flowingdata.com/ppg2008.csv")
nba$Name <- with(nba, reorder(Name, PTS))
nba.m <- melt(nba)
nba.m <- ddply(nba.m, .(variable), transform, rescale = rescale(value))

p <- ggplot(nba.m, aes(variable, Name)) + geom_tile(aes(fill = rescale),
     colour = "white") + scale_fill_gradient(low = "white",
     high = "steelblue")

miRNA_population <- countsPlasma
miRNA_population[miRNA_population>0]<-1
miRNA_population <- tibble::rownames_to_column(miRNA_population) %>%
  rename('rowname' = 'miRNA')
miRNA_population$miRNA <- with(miRNA_population, reorder(miRNA, PAC006))

melt_miRNA <- melt(miRNA_population)

base_size <- 9
ggplot(data = melt_miRNA, aes(x = variable, y = miRNA)) +
  geom_tile(aes(fill=melt_miRNA$value), colour = "white") + 
  scale_fill_gradient(low = "white", high = "blue") + 
  theme_grey(base_size = base_size) + labs(x = "", y = "") + 
  scale_x_discrete(expand = c(0, 0)) + 
  scale_y_discrete(expand = c(0, 0)) +
  ggtitle("miRNA Population Variability\nMaternal Plasma") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))


```

We start with a total of 1571 unique miRNAs - of which no single sample has more than 973 individual miRNAs. 

# Filter sequencing noise   

Here any individual observation with less than 5 counts is considered to be sequencing noise and is reduced to zero and removed by filtering.  
  * Counts less than 5 reduced to 0
  * Logic check and record how many miRs will be filtered
  * Filter to remove miRs with zero counts in all samples

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# exclude any individual count less or equal to than 5
countsPlasma[countsPlasma <= 5] <- 0

# check to establish if any miRNA have no expression across all samples
table(rowSums(countsPlasma == 0) == 48)

# remove miRNAs with zero counts in all samples
countsPlasma <- countsPlasma[ rowSums(countsPlasma) !=0, ]

```

After filtering out the sequencing noise (sequence counts with low confidence) we are left with 1453 unique miRNAs (118 are filtered as counts noise).

# Establish DGEList object

The counts and samples data can now be combined into a single list object for use in differential expression analysis downstream.

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# assign object to DGEList
DGEList_plasma <- DGEList(counts = countsPlasma, samples = samples)

```

# TMM normalisation of libraries

Before differential expression analysis the sample libraries need to be normalised to account for differences in initial library size.     
Normalising the libraries allows for the direct comparison between samples.   
Here the Trimmed Mean of M Values method is used.   
Prior to normalisation the plasma libraries are within a similar range, however the mean of counts is small and centered around -5 CPM (log2)   

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# Calculate and visualise normalisation factors
nsamples <- ncol(DGEList_plasma)
col <- brewer.pal(nsamples, "Paired")

lcpm <- cpm(DGEList_plasma, log=TRUE)
boxplot(lcpm, las = 2, col = col, main = "")
title(main = "Maternal Plasma: Un-normalised data",ylab = "Log-cpm") 

DGEList_plasma <- calcNormFactors(DGEList_plasma, method = "TMM")

```

# Filter biological noise and plot 

Once the libraries are normalised we can filter for biological noise - i.e. miRNAs that are present in such small numbers that they are not biologically relevant. Here a count of less than 1 CPM in 30 or more samples is considered biologically relevant for this analysis. A count of 2 CPM is the equivalant of ~12 transcripts in the smallest library. The number of samples (here 30) is chosen to attempt to overcome the large differences in library sizes. Using a samples cut off lower than 30 fails to adequately change the distribution of counts.   

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# set ggplot theme
theme_set(theme_bw())

# new df of unfiltered cpm for the reduced DGEList
rawCpm_plasma <- cpm(DGEList_plasma)

# new df of unfiltered log 2 cpm for the reduced DGEList
rawlcpm_plasma <- cpm(DGEList_plasma, log = TRUE)

x <- 1 # CPM for filtering
z <- 7 # number of samples for filtering

# remove low expressed miRNAs logical eg cpm >1 in >=30 samples (7 samples in the smallest group)
keep.exprs <- rowSums(rawCpm_plasma > x) >= z
DGEList_plasma <- DGEList_plasma[keep.exprs,, keep.lib.sizes = FALSE]

## The density of log-CPM values for pre-filtered data (A) and post-filtered data (B) are shown for each sample. 
## Dotted vertical lines mark the log-CPM of zero threshold (equivalent to a CPM value of 1) used in the filtering step.

nsamples <- ncol(DGEList_plasma)
col <- brewer.pal(nsamples, "Paired")
par(mfrow = c(1,2))
pdf("unfilteredPlasmaProfile.pdf")
plot(density(rawlcpm_plasma[,1]), col = col[1], lwd = 2, ylim = c(0,0.3), las = 2,
  main = "", xlab = "")
title(main = "A. Unfiltered data", xlab = "Log-cpm") 
abline(v = 0, lty = 3)
for (i in 1:nsamples){
  den <- density(rawlcpm_plasma[,i])
  lines(den$x, den$y, col = col[i], lwd = 2)
}
#legend("topright", legend=samples$samplename, text.col=col, bty = "n")
dev.off()

pdf("filteredPlasmaProfile.pdf")
lcpm <- cpm(DGEList_plasma, log=TRUE)
plot(density(lcpm[,1]), col = col[1], lwd = 2, ylim = c(0,0.3), las = 2,
  main = "", xlab = "")
title(main="B. Filtered data", xlab="Log-cpm") 
abline(v = 0, lty = 3)
for (i in 1:nsamples){
  den <- density(lcpm[,i])
lines(den$x, den$y, col = col[i], lwd = 2)
}
#legend("topright", legend=samples$samplename, text.col=col, bty="n")
dev.off()

par(mfrow=c(1,1))

# Plot filtered library sizes against total miRNA
# count the number of 'non-zero' (ie expressed) miRNAs for each sample after filtering
filteredNonZero <- (DGEList_plasma[[1]]) %>%
  as.data.frame()

filteredNonZero <-  t(numcolwise(nonzero)(filteredNonZero)) %>%
  as.data.frame() %>%
  tibble::rownames_to_column()

#png("robustmiRNAPlasma.png")
ggplot(data=filteredNonZero, aes(x = rowname, y = V1, fill = samples$group)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(x = "Sample Name", y = "Count of Unique miRNAs Identified", title = "Count Robust miRNAs in Plasma") + 
  geom_hline(yintercept = 676,linetype = "dashed") +
  geom_text(aes(20, 700, label = "676 Total miRNA"), size = 4, colour = "black") +
  scale_fill_discrete(name = "Trimester")
#dev.off()

# Distribution of normalised and filtered counts data
lcpm <- cpm(DGEList_plasma, log = TRUE)
boxplot(lcpm, las = 2, col = col, main = "")
title(main="Maternal Plasma: Normalised data",ylab = "Log-cpm")

```

### All plasma mean expression

```{r message=FALSE, error=FALSE, warning=FALSE, cache=TRUE}

# create expression DF for normalised plasma counts
expressionPlasma <- cpm(DGEList_plasma$counts) %>%
  as.data.frame()

# subset the top 20 highly expressed miRs by the sum of expression over all samples.
expressionPlasmaMean <- expressionPlasma[order(rowMeans(expressionPlasma), decreasing = TRUE), ] %>%
  as.data.frame() %>%
  tibble::rownames_to_column("miRNAs") %>%
  as.data.frame() %>%
  dplyr::slice(., 1:nrow(.)) %>% 
  mutate(meanCPM = rowMeans(.[grep("^PAC", names(.))])) %>%
  select(., -matches("PAC"))
expressionPlasmaMean$Tissue <- "Plasma"
# expressionPlasmaTop20$TissueSpecificity <- "unknown"
# expressionPlasmaTop20$PregnancySpecificity <- "unknown"

#pandoc.table(expressionPlasmaMean)

```

### table - top 20 miRNA in Maternal Plasma by mean expression value (CPM)

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# create expression DF for normalised plasma counts
expressionPlasma <- cpm(DGEList_plasma$counts) %>%
  as.data.frame()

# subset the top 20 highly expressed miRs by the sum of expression over all samples.
expressionPlasmaTop20 <- expressionPlasma[order(rowMeans(expressionPlasma), decreasing = TRUE), ] %>%
  as.data.frame() %>%
  tibble::rownames_to_column("miRNAs") %>%
  as.data.frame() %>%
  dplyr::slice(., 1:20) %>% 
  mutate(meanCPM = rowMeans(.[grep("^PAC", names(.))])) %>%
  select(., -matches("PAC"))
expressionPlasmaTop20$Tissue <- "Plasma"
# expressionPlasmaTop20$TissueSpecificity <- "unknown"
# expressionPlasmaTop20$PregnancySpecificity <- "unknown"

pandoc.table(expressionPlasmaTop20)

```

### table - top 100 miRNA in Maternal Plasma by mean expression value (CPM)

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# create expression DF for normalised plasma counts
expressionPlasma <- cpm(DGEList_plasma$counts) %>%
  as.data.frame()

# subset the top 20 highly expressed miRs by the sum of expression over all samples.
expressionPlasmaTop100 <- expressionPlasma[order(rowMeans(expressionPlasma), decreasing = TRUE), ] %>%
  as.data.frame() %>%
  tibble::rownames_to_column("miRNAs") %>%
  as.data.frame() %>%
  dplyr::slice(., 1:100) %>% 
  mutate(meanCPM = rowMeans(.[grep("^PAC", names(.))])) %>%
  select(., -matches("PAC"))
expressionPlasmaTop100$Tissue <- "Plasma"
# expressionPlasmaTop20$TissueSpecificity <- "unknown"
# expressionPlasmaTop20$PregnancySpecificity <- "unknown"

#saveRDS(expressionPlasmaTop100, "plasma_100.rds")

```

## Heatmap of lcpm expression in Maternal Plasma (all robust miRs)

```{r warning=FALSE, error=FALSE, message=FALSE, cache=TRUE}

lcpm <- cpm(DGEList_plasma, log = TRUE)

# df with sample name and weeks only
weeks <- samples[, c("samplename", "gestation")]

# df of lcpm expression by weeks' gestation
expressionWithWeek <- as.data.frame(t(lcpm)) %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'samplename')) %>%
  left_join(., weeks, by = 'samplename') %>%
  select(samplename, gestation, everything()) %>%
  select(., -samplename) %>%
  arrange(gestation) %>%
  t() %>%
  set_colnames(. [1,]) %>%
  as.data.frame() 
expressionWithWeek <- expressionWithWeek[-1, ]

# create heatmap of all robust miRNA using lcpm expression
pheatmap(expressionWithWeek,
         cluster_cols = FALSE, 
         cluster_rows = TRUE,
         legend = TRUE,
         scale = "row",
         main = "log2 CPM Expression of Robust miRNA\nin Maternal Plasma",
         show_rownames = FALSE,
         treeheight_row = 0)

# Heat map with weeks as annotation

annotation <- data.frame(samples[, c(1, 3)]) %>%
  set_colnames(c("rowname", "gestation")) %>%
  as.data.frame() %>%
  tibble::column_to_rownames() 

#pdf("heat_plasma.pdf", height = 11, width = 6)
pheatmap(lcpm, 
         annotation = annotation,
         cluster_cols = T,
         cluster_rows = T,
         main = "log2 CPM Expression of Robust miRNA\nin Maternal Plasma",
         legend = T,
         show_rownames = F,
         show_colnames = F,
         cutree_col = 4,
         treeheight_row = 0)
#dev.off()

```

# Time switching miRs

## unique miRs by week counts filtered to exclude < 2 samples per week showing expression.

This code chunk is written to:   
  - filter any sequencing noise from the initial plasma counts table   
  - replace sample names with gestational age by week   
  - remove any rows (miRs) with fewer than 2 counts in a single week (ie where a miR is expressed in < 2 samples in a week it is considered not to be expressed in that week)  
  - aggregate remaining counts into weeks   
  - produce heat maps of binary expression vs no expression  
nb: this analysis uses the countsPlasma object which has been filtered for sequencing noise (raw counts < 5 -> 0). The counts have not been filtered for low expression, and have not been normalised. I consider this acceptable as the filter for minimum representation (2 samples in any week) has been applied, and the counts are only considered as a binary (on/off) varible such that the actual expression values are not relevant.

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# set ggplot2 theme
theme_set(theme_bw())

# exclude any individual count less than 5 - if not already done
#countsPlasma[countsPlasma < 5] <- 0

sampleWeek <- samples[, c(1, 3)] # select on the samplename and gestation (week) columns from the samples df

# create a new transformed df of un-normalised counts with samplename as the first column
countsByWeek <- t(countsPlasma) %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'samplename'))

countsByWeek <- left_join(sampleWeek, countsByWeek, by = "samplename") # left join the samples and counts data

countsByWeek <- dplyr::arrange(countsByWeek, gestation) # arrange ascending by gestation week

# counts data by week & filter
filtWeek06 <- subset(countsByWeek, gestation == 6) %>%
  t()
filtWeek06 <- filtWeek06[ rowSums(filtWeek06 > 0) >= 2, ]  %>%
  data.frame()
filtWeek06$week06 <- rep(1, nrow(filtWeek06)) # make new column filled with 1s
filtWeek06 <- filtWeek06[c(-1, -2), ] %>%
  subset(., select = week06) %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek07 <- subset(countsByWeek, gestation == 7) %>%
  t()
filtWeek07 <- filtWeek07[ rowSums(filtWeek07 > 0) >= 2, ]  %>%
  data.frame()
filtWeek07$week07 <- rep(1, nrow(filtWeek07)) # make new column
filtWeek07 <- filtWeek07[c(-1, -2), ] %>%
  subset(., select = week07)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek08 <- subset(countsByWeek, gestation == 8) %>%
  t()
filtWeek08 <- filtWeek08[ rowSums(filtWeek08 > 0) >= 2, ]   %>%
  data.frame()
filtWeek08$week08 <- rep(1, nrow(filtWeek08)) # make new column
filtWeek08 <- filtWeek08[c(-1, -2), ] %>%
  subset(., select = week08)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek09 <- subset(countsByWeek, gestation == 9) %>%
  t()
filtWeek09 <- filtWeek09[ rowSums(filtWeek09 > 0) >= 2, ]  %>%
  data.frame()
filtWeek09$week09 <- rep(1, nrow(filtWeek09)) # make new column
filtWeek09 <- filtWeek09[c(-1, -2), ] %>%
  subset(., select = week09)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek10 <- subset(countsByWeek, gestation == 10) %>%
  t()
filtWeek10 <- filtWeek10[ rowSums(filtWeek10 > 0) >= 2, ]   %>%
  data.frame()
filtWeek10$week10 <- rep(1, nrow(filtWeek10)) # make new column
filtWeek10 <- filtWeek10[c(-1, -2), ] %>%
  subset(., select = week10)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek11 <- subset(countsByWeek, gestation == 11) %>%
  t()
filtWeek11 <- filtWeek11[ rowSums(filtWeek11 > 0) >= 2, ]  %>%
  data.frame()
filtWeek11$week11 <- rep(1, nrow(filtWeek11)) # make new column
filtWeek11 <- filtWeek11[c(-1, -2), ] %>%
  subset(., select = week11)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

# Week 12 excluded (for now) from further analysis as it only has one sample #
# filtWeek12 <- subset(countsByWeek, gestation == 12) %>%
#  t()
# filtWeek12 <- filtWeek12[ rowSums(filtWeek12 > 0) >= 2, ]  %>%
#   data.frame()
# filtWeek12$week12 <- rep(1, nrow(filtWeek12)) # make new column
# filtWeek12 <- filtWeek12[c(-1, -2), ] %>%
#   subset(., select = week12)  %>%
#   tibble::rownames_to_column() %>%
#   plyr::rename(c('rowname' = 'miRNA')) %>%
#   as.data.frame()

filtWeek13 <- subset(countsByWeek, gestation == 13) %>%
  t()
filtWeek13 <- filtWeek13[ rowSums(filtWeek13 > 0) >= 2, ]  %>%
  data.frame()
filtWeek13$week13 <- rep(1, nrow(filtWeek13)) # make new column
filtWeek13 <- filtWeek13[c(-1, -2), ] %>%
  subset(., select = week13)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek14 <- subset(countsByWeek, gestation == 14) %>%
  t()
filtWeek14 <- filtWeek14[ rowSums(filtWeek14 > 0) >= 2, ]  %>%
  data.frame()
filtWeek14$week14 <- rep(1, nrow(filtWeek14)) # make new column
filtWeek14 <- filtWeek14[c(-1, -2), ] %>%
  subset(., select = week14)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

filtWeek16 <- subset(countsByWeek, gestation == 16) %>%
  t()
filtWeek16 <- filtWeek16[ rowSums(filtWeek16 > 0) >= 2, ]  %>%
  data.frame()
filtWeek16$week16 <- rep(1, nrow(filtWeek16)) # make new column
filtWeek16 <- filtWeek16[c(-1, -2), ] %>%
  subset(., select = week16)  %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA')) %>%
  as.data.frame()

countsByWeek <- full_join(filtWeek06, filtWeek07, by  = "miRNA") %>%
  as.data.frame() %>%
  full_join(., filtWeek08, by = 'miRNA') %>%
  as.data.frame() %>%
  full_join(., filtWeek09, by = 'miRNA') %>%
  as.data.frame() %>%
  full_join(., filtWeek10, by = 'miRNA') %>%
  as.data.frame() %>%
  full_join(., filtWeek11, by = 'miRNA') %>%
  as.data.frame() %>%
  full_join(., filtWeek13, by = 'miRNA') %>%
  as.data.frame() %>%
  full_join(., filtWeek14, by = 'miRNA') %>%
  as.data.frame() %>%
  full_join(., filtWeek16, by = 'miRNA') %>%
  plyr::rename(c('miRNA' = 'rowname')) %>%
  as.data.frame() 

countsByWeek[is.na(countsByWeek)] <- 0 # convert any NAs to zeros

# Plot and save multiple heat maps for binary expression (all filtered miRNAs)

for(i in 1:9) {
  counts_melt <- melt(countsByWeek[((i*60)-59):(i*60),])

p <- ggplot(counts_melt, aes(variable, rowname)) +
  geom_tile(aes(fill = as.factor(value)), colour = "white") +
  scale_fill_manual(values = c("darksalmon", "forestgreen")) +
  labs(x = "", y = "") +
  ggtitle("miRNA Expression Profile\nin Maternal Plasma") +
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 4)) +
  theme(legend.position = "none") +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_text((counts_melt$rowname), size = 4)) +
  coord_fixed(ratio = 1)

png(paste("plasma_heatmap_", i, ".png", sep = ""), width = 600, height = 500, res = 120)
  print(p)
  dev.off()
}


```

## Single week expression after filtering for >= 2 expression in given week

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# miRs with expression in one week only
w6 <- subset(countsByWeek, countsByWeek$week06 == 1 & (rowSums(countsByWeek[, 3:10]) == 0))
w7 <- subset(countsByWeek, countsByWeek$week07 == 1 & (rowSums(countsByWeek[, c(2, 4:10)]) == 0))
w8 <- subset(countsByWeek, countsByWeek$week08 == 1 & (rowSums(countsByWeek[, c(2:3, 5:10)]) == 0))
w9 <- subset(countsByWeek, countsByWeek$week09 == 1 & (rowSums(countsByWeek[, c(2:4, 6:10)]) == 0))
w10 <- subset(countsByWeek, countsByWeek$week10 == 1 & (rowSums(countsByWeek[, c(2:5, 7:10)]) == 0))
w11 <- subset(countsByWeek, countsByWeek$week11 == 1 & (rowSums(countsByWeek[, c(2:6, 8:10)]) == 0))
w13 <- subset(countsByWeek, countsByWeek$week13 == 1 & (rowSums(countsByWeek[, c(2:7, 9:10)]) == 0))
w14 <- subset(countsByWeek, countsByWeek$week14 == 1 & (rowSums(countsByWeek[, c(2:8, 10)]) == 0))
w16 <- subset(countsByWeek, countsByWeek$week16 == 1 & (rowSums(countsByWeek[, 2:9]) == 0))

singleWeekExpressionPlasma <- bind_rows(w6, w7, w8, w9, w10, w11, w13, w14, w16) %>%
  tibble::column_to_rownames()

# only works before the cluster steps
pheatmap(as.matrix(singleWeekExpressionPlasma), 
         cluster_cols = FALSE,
         cluster_rows = FALSE,
         cellwidth = 15, 
         cellheight = 7,
         legend_breaks = c(1, 0),
         legend_labels = c("On", "Off"),
         color = c("darksalmon", "forestgreen"),
         border_color = "white",
         labels_col = c("Week 6", "Week 7", "Week 8", "Week 9", "Week 10", "Week 11", "Week 12", "Week 13", "Week 14", "Week 16"),
         main="Binary Expression in Maternal Plasma",
         show_rownames = FALSE,
         treeheight_row = 0)

# cluster to obtain order for heatmap (default distance method = euclidean)
clust <- hclust(dist(singleWeekExpressionPlasma))

# order oxygenMediatedExpression by cluster order
singleWeekExpressionPlasma <- singleWeekExpressionPlasma[match(clust$labels, rownames(singleWeekExpressionPlasma)), ] %>%
  tibble::rownames_to_column()

# Heatmap of miRNA in plasma with expression >=2 weeks before or after the 10 week line and nothing on the other side.
singleCounts_melt <- melt(singleWeekExpressionPlasma, id = "rowname") %>%
  plyr::rename(c('rowname' = 'miRNA'))
  
ggplot(singleCounts_melt, aes(variable, miRNA)) +
  geom_tile(aes(fill = as.factor(value)), colour = "white") +
  scale_fill_manual(values = c("darksalmon", "forestgreen")) +
  scale_y_discrete(name = "", limits = rev(clust$labels)) +
  labs(x = "Gestation Week", y = "") +
  ggtitle("Single Week Expression of\n miRNA in Maternal Plasma") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  theme(legend.position = "none") +
  theme(axis.title.y = element_blank(),
      axis.text.y = element_text((singleCounts_melt$rowname), size = 4)) +
  coord_fixed(ratio = .5) 

```

## Hypoxic versus oxygentated expression after filtering for >= 2 expression in given week

There is a change in placental function around the 10 weeks' gestation mark when the placenta infiltrates the maternal blood space and begins to extract oxygen and nutrients from the mother as well as discard waste.   
Here I'm looking for miRs which are specific to either to pre or post infiltration stage. To identify these miRs I've used a filter whereby miRs with expression in at least 2 weeks prior to the week 10 cut off (ie columns 2:5) but nothing after are subset. Then miRs with expression in at least 2 samples post the week 10 cut off (ie columns 7:10) but nothing before the cut off are subset. The two subsetted df are then concatonated to produce a df of only miRs which fit the two filter criteria.

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

left <- subset(countsByWeek, (rowSums(countsByWeek[, 2:5]) > 1 & (rowSums(countsByWeek[, 7:10]) < 1)))

right <- subset(countsByWeek, (rowSums(countsByWeek[, 2:5]) < 1 & (rowSums(countsByWeek[, 7:10]) > 1)))

oxygenMediatedExpressionPlasma <- bind_rows(left, right) %>%
  tibble::column_to_rownames()

# only works before the cluster steps
pheatmap(as.matrix(oxygenMediatedExpressionPlasma), 
         cluster_cols = FALSE,
         cluster_rows = TRUE,
         cellwidth = 15, 
         cellheight = 7,
         legend_breaks = c(1, 0),
         legend_labels = c("On", "Off"),
         color = c("darksalmon", "forestgreen"),
         border_color = "white",
         labels_col = c("Week 6", "Week 7", "Week 8", "Week 9", "Week 10", "Week 11", "Week 12", "Week 13", "Week 14", "Week 16"),
         main="Binary Expression in Maternal Plasma",
         show_rownames = FALSE,
         treeheight_row = 0)

# cluster to obtain order for heatmap (default distance method = euclidean)
clust <- hclust(dist(oxygenMediatedExpressionPlasma))

# order oxygenMediatedExpression by cluster order
oxygenMediatedExpressionPlasma <- oxygenMediatedExpressionPlasma[match(clust$labels, rownames(oxygenMediatedExpressionPlasma)), ] %>%
  tibble::rownames_to_column()

# Heatmap of miRNA in plasma with expression >=2 weeks before or after the 10 week line and nothing on the other side.
hypoxicCounts_melt <- melt(oxygenMediatedExpressionPlasma, id = "rowname") %>%
  plyr::rename(c('rowname' = 'miRNA'))
  
#pdf("heatHypoxicPlasma.pdf", height = 11, width = 6)
ggplot(hypoxicCounts_melt, aes(variable, miRNA)) +
  geom_tile(aes(fill = as.factor(value)), colour = "white") +
  scale_fill_manual(values = c("darksalmon", "forestgreen")) +
  scale_y_discrete(name = "", limits = rev(clust$labels)) +
  labs(x = "Gestation Week", y = "") +
  ggtitle("Gestation Dependent miRNA Expression in Maternal Plasma") +
  theme(plot.title = element_text(hjust = 0.5)) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  theme(legend.position = "none") +
  theme(axis.title.y = element_blank(),
  axis.text.y = element_text(hypoxicCounts_melt$rowname)) +
  geom_vline(xintercept = 5.45, colour = "white") + 
  coord_fixed(ratio = 1)
#dev.off()

```

## unique miRs by fetal sex 

  - The countsPlacsma object has been:
    + filtered for sequencing noise (all counts < 5 reduced to 0)
    + filtered for miRs with 0 counts in all samples (after the < 5 to 0 filter)

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# select on the samplename and gestation (week) columns from the samples df
sampleSex <- samples[, c(1, 9)] 

# make a copy of the counts plasma object to increase filter
sexCountsPlasma <- countsPlasma

# filter any counts < x by replacing with 0
x <- 5
sexCountsPlasma[sexCountsPlasma <= x] <- 0

# remove miRNAs with zero counts in all samples
sexCountsPlasma <- sexCountsPlasma[ rowSums(sexCountsPlasma) !=0, ]

# check to establish if any miRNA have no expression across all samples
table(rowSums(sexCountsPlasma == 0) == 48)

# new df of un-normalised counts (sequencing noise already filtered)
countsBySex <- t(sexCountsPlasma) %>%
  as.data.frame() %>%
  tibble::rownames_to_column() # create a new transformed df of un-normalised counts with samplename as the first column
# left join the samples and counts data
countsBySex <- left_join(sampleSex, countsBySex, by = c("samplename" = "rowname"))

# new df of samplename and sex columns
leading <- countsBySex[, 1:2]
# take out first two columns of samplename and sample sex
countsBySex <- countsBySex[, -c(1, 2)]
# replace all non zeros with one
countsBySex[countsBySex > 0] <- 1
# cbind the samplename and sex back onto the counts
countsBySex <- cbind(leading, countsBySex)

```

# subset expression by fetal sex

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# subset the female samples from counts - may end up with some zero counts
allFemale <- dplyr::filter(countsBySex, Sex == "Female") %>%
  plyr::rename(c('samplename' = 'rowname')) %>% 
  tibble::column_to_rownames() %>%
  select(-matches("Sex")) %>%
  t() %>%
  as.data.frame()

# miRNA must be expressed in min n female samples to be considered biologically relevant in female
n <- 5
biologicallyRelevantFemale <- allFemale[ rowSums(allFemale > 0) >= n, ] %>%
  as.data.frame()

# keep only rows (miRNA) whith zero expression in every sample
zeroExpressionFemale <- allFemale[ rowSums(allFemale) == 0, ] %>%
  as.data.frame()
# pandoc.table(zeroExpressionFemale,
#              caption = "miRNA with zero expression in maternal plasma with female fetus",
#              split.table = Inf)

# keep only rows (miRNA) whith expression (may not be in every sample)
ExpressionFemale <- allFemale[ !rowSums(allFemale) == 0, ] %>%
  as.data.frame()

# subset the male samples from counts - may end up with some zero counts
allMale <- dplyr::filter(countsBySex, Sex == "Male") %>%
  plyr::rename(c('samplename' = 'rowname')) %>% 
  tibble::column_to_rownames() %>%
  select(-matches("Sex")) %>%
  t() %>%
  as.data.frame()
  
# miRNA must be expressed in min n male samples to be considered biologically relevant in male
biologicallyRelevantMale <- allMale[ rowSums(allMale > 0) >= n, ] %>%
  as.data.frame()

# keep only rows (miRNA) whith zero expression in every sample
zeroExpressionMale <- allMale[ rowSums(allMale) == 0, ] %>%
  as.data.frame() 
# pandoc.table(zeroExpressionMale,
#              caption = "miRNA with zero expression in maternal plasma with male fetus",
#              split.table = Inf)

# keep only rows (miRNA) whith expression (may not be in every sample)
ExpressionMale <- allMale[ !rowSums(allMale) == 0, ] %>%
  as.data.frame()

```

# sex specific expression - female

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

zeroMale <- as.data.frame(rownames(zeroExpressionMale)) %>%
  set_colnames("miRNA")

allfemale <- allFemale %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA'))

bioFemale <- as.data.frame(rownames(biologicallyRelevantFemale)) %>%
  set_colnames("miRNA")

femaleOnly <- semi_join(allfemale, zeroMale, by = "miRNA") %>%
  plyr::rename(c('miRNA' = 'rowname')) %>%
  tibble::column_to_rownames()

femaleOnly$numberSamples <- as.data.frame(rowSums(femaleOnly))

# what proportion of the female samples express the 'female specific' miRNA?
proportionFemale <- (femaleOnly$numberSamples)/(ncol(femaleOnly)-1)

(cbind(femaleOnly$numberSamples, proportionFemale) %>%
    set_colnames(c("numberSamples", "proportionSamples")))

# xtable(cbind(femaleOnly$numberSamples, proportionFemale) %>%
#   set_colnames(c("numberSamples", "proportionSamples")))

```

# sex specific expression - male

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

zeroFem <- as.data.frame(rownames(zeroExpressionFemale)) %>%
  set_colnames("miRNA")

allmale <- allMale %>%
  tibble::rownames_to_column() %>%
  plyr::rename(c('rowname' = 'miRNA'))

bioMale <- as.data.frame(rownames(biologicallyRelevantMale)) %>%
  set_colnames("miRNA")

maleOnly <- semi_join(allmale, zeroFem, by = "miRNA") %>%
  plyr::rename(c('miRNA' = 'rowname')) %>%
  tibble::column_to_rownames()

# How anyof the male samples express the 'male specific' miRNA
maleOnly$numberSamples <- as.data.frame(rowSums(maleOnly)) 

# what proportion of the male samples express the 'male specific' miRNA?
proportionMale <- (maleOnly$numberSamples)/(ncol(maleOnly)-1)

cbind(maleOnly$numberSamples, proportionMale) %>%
  set_colnames(c("numberSamples", "proportionSamples"))

# xtable(cbind(maleOnly$numberSamples, proportionMale) %>%
#   set_colnames(c("numberSamples", "proportionSamples")))

```

# CPM lists for correlation exploration

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# create object holding the normalised counts per million for plasma samples and save
cpmPlasmaNormal <- cpm(DGEList_plasma)

#saveRDS(cpmPlasmaNormal, "cpmPlasmaNormal.rds")

```

# Looking for Hemolysis - miRNA associated with hemolysis
  - I've split the miRNAs associated with hemolysis from the Kirschner 2013 paper into two sets
    - low expression set: expression values tend to be in the 100s
    - high expression set: expression values are in 1000s or 10,000s
  - I've plotted these seperately because it was difficult to see the expression changes in the low expression miRNAs

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# miRNA to asses ratio of stable versus erythrocyte associated molecules from Blondal et al 2012
plasma_miR <- c("hsa-miR-451a", "hsa-miR-23a-3p")

# miRNA to asses expression of hemolysis associated molecules from Kirschner et al 2013
plasma_miR_low <- c("hsa-miR-15b-3p", "hsa-miR-15b-5p", "hsa-miR-486-3p", "hsa-miR-532-3p", "hsa-miR-532-5p", "hsa-miR-636", "hsa-miR-1255b-5p")

plasma_miR_high <- c("hsa-miR-16-5p", "hsa-miR-451a", "hsa-miR-486-5p")

# gestational order for plots
gestation_order <- sampleWeek %>%
  arrange(gestation)

###### CPM CPM CPM CPM CPM ########
cpm <- cpm(DGEList_plasma$counts)

# subset the miRNA of interest in hemolysis analysis
cpm_hemoCountsPlasma <- subset(cpm, rownames(cpm) %in% plasma_miR) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  set_colnames(c("samplename", "23a", "451a"))

# add columns for difference and ratio of miRs of interest
cpm_hemoCountsPlasma <- cpm_hemoCountsPlasma %>%
  as.tibble() %>%
  dplyr::mutate(delta_cpm = `23a`-`451a`,
         ratio_cpm = `23a`/`451a`) %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename')

#subset the miRNA of interest in hemolysis analysis
cpm_hemoCountsPlasma_low <- subset(cpm, rownames(cpm) %in% plasma_miR_low) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  rename(., 'rowname' = 'samplename')

# add columns for difference and ratio of miRs of interest
cpm_hemoCountsPlasma_low <- cpm_hemoCountsPlasma_low %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename') %>%
  arrange(gestation)

#subset the miRNA of interest in hemolysis analysis
cpm_hemoCountsPlasma_high <- subset(cpm, rownames(cpm) %in% plasma_miR_high) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  rename(., 'rowname' = 'samplename')

# add columns for difference and ratio of miRs of interest
cpm_hemoCountsPlasma_high <- cpm_hemoCountsPlasma_high %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename') %>%
  arrange(gestation)

###### PLOT PLOT PLOT PLOT PLOT PLOT PLOT ########

ggplot(data = cpm_hemoCountsPlasma, aes(x = `23a`, y = `451a`)) +
  geom_point() +
  geom_line() +
  theme_bw()

cpm_hemoCountsPlasma <- cpm_hemoCountsPlasma %>%
  arrange(desc(delta_cpm))

ggplot(data = cpm_hemoCountsPlasma, aes(x = samplename, y = delta_cpm, fill = gestation)) +
  geom_bar(stat="identity") +
  scale_x_discrete(limits=cpm_hemoCountsPlasma$samplename) +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))

cpm_hemoCountsPlasma <- cpm_hemoCountsPlasma %>%
  arrange(desc(ratio_cpm))

ggplot(data = cpm_hemoCountsPlasma, aes(x = samplename, y = ratio_cpm, fill = gestation)) +
  geom_bar(stat="identity") +
  scale_x_discrete(limits=cpm_hemoCountsPlasma$samplename) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))

# melt the dataset into long for so I can stagger the bars in the bar plot
cpm_hemoCountsPlasma_low_melt <- melt(cpm_hemoCountsPlasma_low, 
                     variable.name = "miRNA",
                     value.names = "value",
                     id.vars = c("samplename", "gestation"))

# bar plot with each miRNA as a seperate vertical bar against each sample
ggplot(data = cpm_hemoCountsPlasma_low_melt, 
       aes(x = factor(samplename, level = gestation_order$samplename), 
           y = value, fill = miRNA)) +
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_brewer(palette = "Spectral") +
  ggtitle("Expression of Hemolysis Associated \nmiRNA in Maternal Plasma (low)") +
  xlab("Samples by ascending gestation") +
  ylab("Normalised CPM Expression") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))

# melt the dataset into long for so I can stagger the bars in the bar plot
cpm_hemoCountsPlasma_high_melt <- melt(cpm_hemoCountsPlasma_high, 
                     variable.name = "miRNA",
                     value.names = "value",
                     id.vars = c("samplename", "gestation"))

# bar plot with each miRNA as a seperate vertical bar against each sample
ggplot(data = cpm_hemoCountsPlasma_high_melt, 
       aes(x = factor(samplename, level = gestation_order$samplename), 
           y = value, fill = miRNA)) +
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_manual(values = c("red", "green", "blue")) +
  ggtitle("Expression of Hemolysis Associated \nmiRNA in Maternal Plasma (high)") +
  xlab("Samples by ascending gestation") +
  ylab("Normalised CPM Expression") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))


# ###### logCPM logCPM logCPM logCPM logCPM ########
# 
# # subset the miRNA of interest in hemolysis analysis
# lcpm_hemoCountsPlasma <- subset(lcpm, rownames(lcpm) %in% plasma_miR) %>%
#   t() %>%
#   as.data.frame() %>%
#   tibble::rownames_to_column() %>%
#   set_colnames(c("rowname", "23a", "451a"))
# 
# # add columns for difference and ratio of miRs of interest
# lcpm_hemoCountsPlasma <- lcpm_hemoCountsPlasma %>%
#   as.tibble() %>%
#   dplyr::mutate(delta_lcpm = `23a`-`451a`,
#          ratio_lcpm = `23a`/`451a`) %>%
#   as.data.frame()
# 
# ###### PLOT PLOT PLOT PLOT PLOT PLOT PLOT ########
# 
# ggplot(data = lcpm_hemoCountsPlasma, aes(x = `23a`, y = `451a`)) +
#   geom_point() +
#   theme_bw()
# 
# lcpm_hemoCountsPlasma <- lcpm_hemoCountsPlasma %>% 
#   arrange(desc(delta_lcpm))
# 
# ggplot(data = lcpm_hemoCountsPlasma, aes(x = rowname, y = delta_lcpm)) +
#   geom_bar(stat="identity") +
#   scale_x_discrete(limits=lcpm_hemoCountsPlasma$rowname) +
#   theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))


```

# Looking for Hemolysis - miRNA unaffected by hemolysis

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}


# miRNA to asses expression of stable molecules from Kirschner et al 2013
stable_miR_low <- c("hsa-miR-1274b", "hsa-miR-142-3p", "hsa-miR-142-5p", "hsa-miR-146a-3p", "hsa-miR-146a-5p", "hsa-miR-122-3p", "hsa-miR-122-5p", "hsa-miR-23a-3p")

###### CPM CPM CPM CPM CPM ########
#cpm <- cpm(DGEList_plasma$counts)

#subset the miRNA of interest in hemolysis analysis
cpm_stableCountsPlasma <- subset(cpm, rownames(cpm) %in% stable_miR_low) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  rename(., 'rowname' = 'samplename')

# add columns for difference and ratio of miRs of interest
cpm_stableCountsPlasma <- cpm_stableCountsPlasma %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename') %>%
  arrange(gestation)

###### PLOT PLOT PLOT PLOT PLOT PLOT PLOT ########

# melt the dataset into long for so I can stagger the bars in the bar plot
cpm_stableCountsPlasma_melt <- melt(cpm_stableCountsPlasma, 
                     variable.name = "miRNA",
                     value.names = "value",
                     id.vars = c("samplename", "gestation"))

# bar plot with each miRNA as a seperate vertical bar against each sample
ggplot(data = cpm_stableCountsPlasma_melt, 
       aes(x = factor(samplename, level = gestation_order$samplename), 
           y = value, fill = miRNA)) +
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_brewer(palette = "Spectral") +
  ggtitle("Expression of Hemolysis Stable \nmiRNA in Maternal Plasma") +
  xlab("Samples by ascending gestation") +
  ylab("Normalised CPM Expression") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))

```

# Looking for Hemolysis - normalise miR-451a using miR-23a-3p expression
  - miR-23a is known to be stable in plasma before and after hemolysis
  By normalising both miR-23a and miR-451a I plan to look at the relative expression of these two transcripts in each sample on a 'per sample' basis.

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}

# miRNA to asses expression of stable molecules from Kirschner et al 2013
stable_miR_ratio_451a <- c("hsa-miR-23a-3p", "hsa-miR-451a")

# subset the miRNA of interest in hemolysis analysis
cpm_hemoCountsPlasma_stable_451a <- subset(cpm, rownames(cpm) %in% stable_miR_ratio_451a) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  set_colnames(c("samplename", "23a", "451a"))

# add columns for difference and ratio of miRs of interest
cpm_hemoCountsPlasma_stable_451a <- cpm_hemoCountsPlasma_stable_451a %>%
  as.tibble() %>%
  dplyr::mutate(normalised_23a = `23a`/`23a`,
         normalised_451a = `451a`/`23a`) %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename')

ggplot(data = cpm_hemoCountsPlasma_stable_451a, aes(x = samplename, y = normalised_451a)) +
  geom_bar(stat="identity") +
  scale_x_discrete(limits=cpm_hemoCountsPlasma_stable_451a$samplename) +
  geom_hline(yintercept=1,linetype = "dashed", color = "blue") +
  geom_hline(yintercept=32,linetype = "dashed", color = "green") +
  geom_hline(yintercept=128,linetype = "dashed", color = "red") +
  geom_hline(yintercept=256,linetype = "dashed", color = "red") +
  ggtitle("Sequencing approximation of delta Cq \nmiR-451a/miR-23a") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))


```

# Looking for Hemolysis - normalise miR-451a using miR-23a-3p expression on a "per sample" basis where:
  - potential pseudo Cq curves are plotted assuming the copy number for miR-23a-3p starts at 1.
  - both curves use the FCN = ICN * 2^# cycles algorithm 

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}


# miRNA to asses expression of stable molecules from Kirschner et al 2013
stable_miR_ratio_451a <- c("hsa-miR-23a-3p", "hsa-miR-451a")

# subset the miRNA of interest in hemolysis analysis
cpm_hemoCountsPlasma_stable_451a <- subset(cpm, rownames(cpm) %in% stable_miR_ratio_451a) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  set_colnames(c("samplename", "miR_23a", "miR_451a"))

# add columns for difference and ratio of miRs of interest
cpm_hemoCountsPlasma_stable_451a <- cpm_hemoCountsPlasma_stable_451a %>%
  as.tibble() %>%
  dplyr::mutate(normalised_23a = `miR_23a`/`miR_23a`,
         normalised_451a = `miR_451a`/`miR_23a`) %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename') %>%
  filter(., samplename == "PAC017")

test <- rbind(cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a, cpm_hemoCountsPlasma_stable_451a)
test <- test[, c(1:5)]
cycle <- as.data.frame(c(0:8)) %>%
  set_colnames("cycle")

test <- cbind(test, cycle)
test <- test %>%
  dplyr::mutate(Cq_23a = normalised_23a*2^cycle,
                Cq_451a = normalised_451a*2^cycle)

ggplot(data = test, aes(x = cycle, y=Cq_23a)) +
  geom_point()

test_melt <- test[, c(1, 6:8)] %>%
  melt(., id = c("samplename", "cycle"))

ggplot(data = test_melt, aes(x = cycle, y=value, fill=variable)) +
  geom_point() + 
  geom_line() +
  ggtitle("PAC017") +
  theme_bw()

```

# Looking for Hemolysis - normalise miR-16-5p using miR-23a-3p expression
  - miR-23a is known to be stable in plasma before and after hemolysis
  By normalising both miR-23a and miR-451a I plan to look at the relative expression of these two transcripts in each sample on a 'per sample' basis.

```{r message=FALSE, warning=FALSE, error=FALSE, cache=TRUE}


# miRNA to asses expression of stable molecules from Kirschner et al 2013
stable_miR_ratio_16 <- c("hsa-miR-23a-3p", "hsa-miR-16-5p")

# subset the miRNA of interest in hemolysis analysis
cpm_hemoCountsPlasma_stable_16 <- subset(cpm, rownames(cpm) %in% stable_miR_ratio_16) %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column() %>%
  set_colnames(c("samplename", "miR_23a", "miR_16"))

# add columns for difference and ratio of miRs of interest
cpm_hemoCountsPlasma_stable_16 <- cpm_hemoCountsPlasma_stable_16 %>%
  as.tibble() %>%
  dplyr::mutate(normalised_23a = `miR_23a`/`miR_23a`,
         normalised_16 = `miR_16`/`miR_23a`,
         normalised_23a = `miR_23a`/`miR_16`) %>%
  as.data.frame() %>%
  left_join(., sampleWeek, by='samplename')

ggplot(data = cpm_hemoCountsPlasma_stable_16, aes(x = samplename, y = normalised_23a)) +
  geom_bar(stat="identity") +
  scale_x_discrete(limits=cpm_hemoCountsPlasma_stable_16$samplename) +
  ggtitle("Sequencing approximation of delta Cq \nmiR-16-5p/miR-23a") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust=1, vjust=1))


```

```{r echo=FALSE}

today <- Sys.Date()

```

This R session was run on `r format(today, format="%B %d %Y")`.

```{r}

sessionInfo()

```



